tests/test_repro.py

```diff
@@ -53,6 +54,138 @@ def test(self):
         self.assertNotEqual(ret, 0)
 
 
+class TestReproCyclicGraph(TestDvc):
+    def test(self):
+        self.dvc.run(deps=[self.FOO],
+                     outs=['bar.txt'],
+                     cmd='echo bar > bar.txt')
+
+        self.dvc.run(deps=['bar.txt'],
+                     outs=['baz.txt'],
+                     cmd='echo baz > baz.txt')
+
+        with open('cycle.dvc', 'w') as fd:
+            stage_dump = {
+                'cmd': 'echo baz > foo',
+                'deps': [{'path': 'baz.txt'}],
+                'outs': [{'path': self.FOO}],
+            }
+
+            yaml.safe_dump(stage_dump, fd, default_flow_style=False)
+
+        with self.assertRaises(CyclicGraphError):
+            self.dvc.reproduce('cycle.dvc')
+
+
+class TestReproWorkingDirectoryAsOutput(TestDvc):
+    """
+    |  stage.cwd  |  out.path | cwd as output |
+    |:-----------:|:---------:|:-------------:|
+    |     dir     |    dir    |      True     |
+    | dir/subdir/ |    dir    |      True     |
+    |     dir     |   dir-1   |     False     |
+    |      .      | something |     False     |
+    """
+    def test(self):
+        # File structure:
+        #       .
+        #       ├── dir1
+        #       │  └── dir2.dvc         (out.path == ../dir2)
+        #       └── dir2
+        #          └── something.dvc    (stage.cwd == ./dir2)
+
+        os.mkdir(os.path.join(self.dvc.root_dir, 'dir1'))
+
+        self.dvc.run(
+            cwd='dir1',
+            outs=['../dir2'],
+            cmd='mkdir {path}'.format(path=os.path.join('..', 'dir2'))
+        )
+
+        faulty_stage_path = os.path.join('dir2', 'something.dvc')
+
+        with open(faulty_stage_path, 'w') as fd:
+            stage_dump = {
+                'cmd': 'echo something > something',
+                'outs': [{'path': 'something'}],
+            }
+
+            yaml.safe_dump(stage_dump, fd, default_flow_style=False)
+
+        with self.assertRaises(WorkingDirectoryAsOutputError):
+            self.dvc.reproduce(faulty_stage_path)
+
+    def test_nested(self):
```

Comment:
Well, looks like the order of walk() is not guaranteed to be the same :) I see it failing on CI and not failing when I run them on my machine, so I guess we need to mock Project.stages() to return both `error.dvc` and `nested.dvc` . Could you please submit a change a for that?